### ISSUE TITLE
Work on PyTorch Python code without having to build from source

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#40829]
One of the biggest barriers to contributing small patches to PyTorch is inability to get a working development environment, because doing a full PyTorch build can take hours if you do not have very many CPUs. This is a massive waste if you just want to work on Python code, since you aren't going to need to rebuild anything in this case, why do you have to wait hours just to get started on your patch?

Can we provide *precompiled* binaries for PyTorch developers who do not need to rebuild C++? The answer is yes! We have nightlies, which give you fairly up-to-date compilation products. To take an example, I downloaded the latest OS X nightly:

```
conda install -c pytorch-nightly pytorch
```

then, in a separate conda environment, I installed the required dependencies:

```
conda install --only-deps -c pytorch-nightly pytorch
```

and checked out a PyTorch source copy for the same hash as the nightly (this was identified by looking at `torch.version.git_version` of the nightly, and then cross-referencing that hash with the actual master hash in the description)

```
>>> torch.version.git_version
'28d1a1436bd51092843392025f184b1f428745b2'

$ git show 28d1a1436bd51092843392025f184b1f428745b2
commit 28d1a1436bd51092843392025f184b1f428745b2
Author: chronos_secgrp_pytorch_oss_ci_oncall <chronos_secgrp_pytorch_oss_ci_oncall@chronos500.prn2.facebook.com>
Date:   Fri Jun 26 00:02:56 2020 -0700

    2020-06-26 nightly release (eebd492dcfb8d4152a5ba4240486f3df5aaf112d)

$ git checkout eebd492dcfb8d4152a5ba4240486f3df5aaf112d
```

Then inside the install directory of the nightly `/Users/ezyang/miniconda3/lib/python3.7/site-packages/torch` of the nightly, copied the relevant binaries to my source directory

```
cp *.so ~/Dev/pytorch-tmp/torch
cp -R lib ~/Dev/pytorch-tmp/torch
cp version.py ~/Dev/pytorch-tmp/torch/
cp -R bin ~/Dev/pytorch-tmp/torch/
```

This got me a 95% working version of PyTorch (the rpath on the shm manager was wrong so shared memory management didn't work).

It would be great if we could have a script that automated this process in a cross-platform way, so people can easily check out a warmed copy of PyTorch corresponding to a nightly and then hack away.

cc @ezyang @seemethere @malfet @rgommers

**Comment by scopatz** [comment]
It seems like it would be better to `conda install` into the target directory, rather than copying over and having to deal with weird linkage issues. Sinc this is suppossed to be cross-platform, would a python script be good enough? Do you want a `conda pytorch dev` command or something similar?

**Comment by rgommers** [comment]
> Do you want a `conda pytorch dev` command or something similar?

Sounds cool. Is it possible to make sure that that command doesn't overwrite files that are under version control?

**Comment by scopatz** [comment]
Yep! That is doable.

**Comment by scopatz** [comment]
@ezyang - What where you doing exactly to test that this was working? 

Also the straight `conda install` into the git repo probably won't work because the structure of the package tarball doesn't match the structure of the repo.

**Comment by ezyang** [comment]
> It seems like it would be better to conda install into the target directory, rather than copying over and having to deal with weird linkage issues. Since this is suppossed to be cross-platform, would a python script be good enough? Do you want a conda pytorch dev command or something similar?

You definitely don't have to do it the way I tried to hack it up, I just mostly wanted to see if it was feasible. A Python script seems good.

>  What where you doing exactly to test that this was working?

Things I'm most interested in enabling:

* Working on docs locally without having to build
* Working on type annotations without having to build
* Maybe simple work on Python tests (but nothing too fancy; since most core functionality changes would require you to actually rebuild)

**Comment by scopatz** [comment]
OK - I have a basic version for linux working (I think). I'll flesh out the two functions that need it for mac and windows tomorrow.  Then hopefully, you can test it!

**Comment by scopatz** [comment]
Implementation available in #42635
